### PR TITLE
feat(agent): heartbeat & error handling with audit log

### PR DESCRIPTION
**Pull request by unclesp1d3r**
## Summary

Implements [Agent Heartbeat & Error Handling](../blob/main/spec/tickets/Agent_Heartbeat_&_Error_Handling.md) for the hashcat worker agent surface.

## Implementation units

- **U1** — `agentHeartbeatSchema` + `agentHeartbeatErrorSchema` + `agentHeartbeatCurrentTaskSchema` in `@hashhive/shared`; corresponding `AgentHeartbeatError` / `AgentHeartbeatCurrentTask` type exports.
- **U2/U3** — `processHeartbeat` refactored with a pure `decideHeartbeatTransition` decision helper, transactional persistence, ownership-checked task linkage, structured status-transition audit log, and post-commit event emission.
- **U4** — Route handler at `POST /api/v1/agent/heartbeat` flows the new schema through unchanged.
- **U5** — Pure-helper tests, schema-validation tests, and full integration coverage for the heartbeat path.

## Hardening applied during review (commits `ca6a144` -> `4c94b281e78d53d6c74cb98f92d97d66e26bb485`)

Every finding from the multi-agent review pass + CodeRabbit + the second-pass pr-review-toolkit run is addressed in-PR:

### Schema + DoS bounds
- Bound `error.message` at 4096 chars and `error.context` at 16 K serialized chars on both heartbeat and standalone `/errors` schemas.
- Documented size caps and the narrow `warning|fatal` enum on the OpenAPI spec.

### Logging + visibility
- Structured `logger.info` audit line on every real status transition; suppressed on no-op heartbeats.
- `logger.warn` when a heartbeat arrives for a vanished agent row.
- `logger.error` when `agent_errors` INSERT returns no row.
- `logger.warn` when `currentTask.taskId` is not owned by the calling agent (drops the task linkage).
- `logger.error` when `handleTaskFailure` throws mid-loop (continue-on-error preserves sibling tasks).

### Security
- PII scrubber (`scrubAgentErrorContext`) redacts `/token|password|secret|api[_-]?key|authorization|cookie|bearer/i` keys server-side before persisting.
- `currentTask.taskId` is verified against `tasks WHERE id=? AND agent_id=?` before persistence so a compromised agent token cannot attribute errors to another agent's tasks.
- FK on `agent_errors.task_id` (ON DELETE SET NULL) as last-line guard; migration 0007 backfills orphan refs before attaching.
- Fatal-heartbeat recovery: split agent-token middleware into strict (work endpoints) and recovery-friendly (`requireAgentTokenAllowRecovery`, heartbeat-only) variants so an errored agent can self-recover via a clean heartbeat.

### Atomicity + correctness
- `db.transaction` wraps the heartbeat-error INSERT + agents UPDATE; `SELECT ... FOR UPDATE` on the agent row closes the TOCTOU race on prior-status read.
- Event emits (`emitAgentError`, `emitAgentStatus`) fire AFTER the transaction commits so SSE clients never see rolled-back state.
- Fatal task-failure loop wraps each `handleTaskFailure` in try/catch; partial failures surface in the response as `{attempted, failed}`.
- `HeartbeatTransition` refactored to a discriminated union (`kind: 'noop' | 'transition'`); invalid combinations are compile errors and the call-site triple guard collapses to one kind check.

### Type tightening
- `decideHeartbeatTransition` parameter and `HeartbeatTransition.effectiveStatus` use the `AgentHeartbeat['status']` literal union — status typos at call sites are compile errors.
- `errorSeverity` parameter typed via `AgentHeartbeatError['severity']`.

### Contract + ergonomics
- All agent-API zValidator failures return the documented `{error: {code: 'VALIDATION_ERROR', message}}` envelope via a centralized `agentValidationHook`.
- Mock-leak fix: contract test re-exports the real `decideHeartbeatTransition`, `classifyWorstSeverity`, `classifyRecentErrors`, `pickCurrentTaskByAgent`, and `scrubAgentErrorContext` from its `mock.module` factory to avoid Linux-vs-macOS load-order CI flakes (GOTCHAS.md pattern).
- `logAgentError` accepts an optional drizzle client (so it can participate in an outer transaction) and an optional `projectId` pass-through (so `processHeartbeat` skips the redundant SELECT).

### OpenAPI spec
- `Heartbeat` schema documents `currentTask`, `error`, the narrow severity enum, size caps, the non-redundant heartbeat-vs-`/errors` contract, and the recovery path.

## Test coverage

| Phase | Tests | Notes |
|-------|-------|-------|
| Main suite | 372 pass / 0 fail / 15 skip | Unit + route contract |
| Heartbeat integration (`AGENT_HEARTBEAT_TEST_ISOLATED=1`) | 9 pass | Full `processHeartbeat` path with mocked drizzle; 1:1 mapped to plan U5 |
| Tasks isolated | 22 pass | Existing |
| Queue-manager isolated | 19 pass | Existing |
| Control-RBAC isolated | 3 pass | Existing |
| Health-deterministic isolated | 9 pass | Existing |

Plan U5 scenarios covered by the new `tests/integration/agent-heartbeat.test.ts`:
- warning persistence (R3/R4)
- fatal → `handleTaskFailure` invocation (R3/R4/R5)
- fatal with no active tasks (R5 edge)
- transition log fires once on real transition (R2)
- no transition log on no-op heartbeat (R2)
- `currentTask.taskId` carryover when owned
- `currentTask.taskId` dropped when not owned (with audit log)
- mid-loop `handleTaskFailure` throw doesn't strand siblings
- `error.context` PII scrub before persistence

## Test plan

- [x] `just check` (format, lint, type-check, build) — clean
- [x] `just ci-check` (full test suite on top of check) — 372 main + 9 heartbeat integration + 53 other isolated = 396 total, 0 fail, 15 skip
- [x] Plan U5 integration tests — covered (see table above)
- [x] Manual exercise against a local dev server — heartbeat with no error / warning / fatal / malformed all behave as specified

## Refs

- Spec: `spec/tickets/Agent_Heartbeat_&_Error_Handling.md`
- Plan: `docs/plans/2026-05-16-001-feat-agent-heartbeat-error-handling-plan.md` (gitignored)
- Migration: `packages/shared/src/db/migrations/0007_light_ulik.sql`

### CODE REVIEW

File: packages/backend/tests/unit/agents-service.test.ts [hunk 1]
```diff
@@ -267,3 +268,70 @@ describe('AGENT_TASK_ACTIVE_STATUSES', () => {
...
+  });
+
+  it('emits a fatal_error transition when fatal flips an agent from online', () => {
+    const result = decideHeartbeatTransition({
+      payloadStatus: 'online',
+      errorSeverity: 'fatal',
+      priorStatus: 'online',
+    });
+    expect(result.shouldLogTransition).toBe(true);
+    expect(result.reason).toBe('fatal_error');
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/unit/agents-service.test.ts]: > The second one (lines 319-327) is redundant with the first (lines 273-283) and adds no additional coverage — consider replacing it with a scenario that actually differs (e.g., fatal while `priorStatus` is already `'error'`, where `effectiveStatus === priorStatus` and `shouldLogTransition` becomes `false`, which is currently untested and is a meaningful edge case in the policy).

Addressed: replaced the redundant case with the suggested edge case in commit e5074af — fatal heartbeat against an already-errored agent now exercises the noop arm of the discriminated union (effectiveStatus stays 'error', isFatalError=true, kind='noop'). See `tests/unit/agents-service.test.ts` 'returns kind=noop on a fatal heartbeat when the agent is already in error state'.

File: packages/backend/src/services/agents.ts [hunk 2]
```diff
@@ -265,20 +272,118 @@ async function fetchCurrentTasks(
...
+      agentId,
+      severity: data.error.severity,
+      message: data.error.message,
+      context: data.error.context,
+      taskId: data.currentTask?.taskId,
+      // Skip the redundant SELECT inside logAgentError — we already
+      // have projectId from priorRow.
+      projectId: priorRow?.projectId,
+    });
   }
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts]: > `data.currentTask?.taskId` from the heartbeat payload is persisted directly into `agent_errors.task_id` with no check that the referenced task is actually owned by the calling agent

Addressed: `verifyTaskOwnership` (in `services/agents.ts`) now queries `tasks WHERE id=? AND agent_id=?` and drops the task linkage on mismatch (logging a `logger.warn` so compromised tokens are detectable). As of commit e5074af the verification runs inside the heartbeat transaction with `FOR UPDATE` on the task row, closing the TOCTOU window between verify and insert. There is also now an FK on `agent_errors.task_id` with `ON DELETE SET NULL` as a last-line guard (migration 0007).

File: packages/backend/src/services/agents.ts [hunk 3]
```diff
@@ -265,20 +272,118 @@ async function fetchCurrentTasks(
...
+    await logAgentError({
+      agentId,
+      severity: data.error.severity,
+      message: data.error.message,
+      context: data.error.context,
+      taskId: data.currentTask?.taskId,
+      // Skip the redundant SELECT inside logAgentError — we already
+      // have projectId from priorRow.
+      projectId: priorRow?.projectId,
+    });
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts]: > `error.context` is persisted verbatim to `agent_errors.context` ... Consider a redaction pass (regex-based key match against `/token|password|secret|api[_-]?key|authorization|cookie|bearer/i`) before insert

Addressed: `scrubAgentErrorContext` walks the payload (depth-capped at 6) and redacts secret-shaped keys to `[REDACTED]` before persisting. The match is whole-word + trailing-suffix-aware (`isSecretKey`) so legitimate keys like `tokenCount`, `cookieDomain`, `bearerHostname`, and `apiKeyName` are preserved, while compound names like `db_password` and `x_api_key` are redacted. Positive and negative tests pin the boundary.

File: packages/backend/tests/unit/agent-api-contract.test.ts [hunk 4]
```diff
@@ -159,6 +195,217 @@ describe('Agent API: POST /heartbeat', () => {
...
+        status: 'error',
+        error: { severity: 'fatal', message: 'hashcat crashed' },
+      }),
+    });
+
+    // Assert
+    expect(res.status).toBe(200);
+    const body = (await res.json()) as Record<string, unknown>;
+    expect(body['acknowledged']).toBe(true);
+  });
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/unit/agent-api-contract.test.ts:268]: > Add an integration test file (with a test DB, no service mocks) covering each scenario.

Addressed: added `packages/backend/tests/integration/agent-heartbeat.test.ts` with 9 scenarios mapped 1:1 to plan U5 — warning persistence, fatal→handleTaskFailure dispatch, fatal-with-no-active-tasks, transition-log fires once, no-op suppression, `currentTask.taskId` carryover, ownership rejection, mid-loop continue-on-error, and PII scrub. The file runs in its own bun:test phase (`AGENT_HEARTBEAT_TEST_ISOLATED=1`) per the GOTCHAS.md mock-leakage pattern. A real-DB integration is desirable as a follow-up but the project's existing tests/integration/ suite also mocks the drizzle client, so the suite follows the established convention.

File: packages/backend/src/routes/agent/index.ts [hunk 5]
```diff
@@ -25,6 +29,26 @@ import {
...
+function agentValidationHook(
+  result: { success: boolean; error?: { issues?: ZodIssue[] } },
+  c: Context
+) {
+  if (result.success) return;
+  const first = result.error?.issues?.[0];
+  const path = first?.path?.length ? first.path.join('.') : 'body';
+  const message = first?.message ? `${path}: ${first.message}` : 'Invalid request body';
+  return c.json({ error: { code: 'VALIDATION_ERROR', message } }, 400);
+}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/routes/agent/index.ts:71]: > The inline `result` type omits required properties (`data`, `target`) and misdescribes `error` as `{ issues?: ZodIssue[] }` instead of importing `ZodError`. ... Replace with the correct signature from `@hono/zod-validator` 0.7.6

Addressed in commit e5074af: `agentValidationHook` now accepts the documented Hook union (`{ success: true; data; target } | { success: false; error: ZodErrorLike; data; target }`) so `result.success === true` narrows correctly, and `data`/`target` are present in the type. The hook also now walks every issue via a new `formatValidationMessage` helper, joining them with `; ` so union-refinement failures surface in the 400 body instead of being dropped. `ZodErrorLike` is a structural subset of `ZodError` (the project's strict-mode tsc treats `PropertyKey[]` as the issue-path type, which the structural shape captures); using the official generic `Hook<T, E, P, Target, O, Schema>` would re-parameterize the hook per call site, which buys narrowing we don't use here.

File: packages/backend/src/routes/agent/index.ts [hunk 6]
```diff
@@ -25,11 +29,37 @@ import {
...
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/routes/agent/index.ts:71]: > Consider using the typed hook signature `(result, c) => { if (!result.success) ... }` and walking all issues for a clearer 400 body.

Addressed in commit e5074af alongside the related CodeRabbit thread on the same line. The hook is now structurally typed against the @hono/zod-validator 0.7.6 result union (`success: true | false` discriminant, `data`, `target` all present), and the new `formatValidationMessage` walks every issue — so union-refinement and nested-issue messages all reach the 400 body.

File: packages/backend/tests/unit/auth-middleware.test.ts [hunk 7]
```diff
@@ -230,3 +244,55 @@ describe('requireAgentToken middleware', () => {
...
+    mockAgentResult = [{ id: 42, projectId: 7, status: 'online', capabilities: { gpu: true } }];
+
+    // Act
+    const res = await app.request('/agent-endpoint', {
+      headers: { authorization: 'Bearer valid-agent-token' },
+    });
+
+    // Assert
+    expect(res.status).toBe(200);
+  });
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/unit/auth-middleware.test.ts:266]: > Line 218 mock data uses `'active'` which violates the schema contract. Update to `status: 'online'` to match the valid enum and the correct usage at line 257.

Addressed in commit e5074af: `mockAgentResult` at line 218 now uses `status: 'online'`, matching the agent status enum (`online | busy | error | benchmarked`) and the recovery-middleware test at line 257.

File: packages/backend/src/services/agents.ts [hunk 8]
```diff
@@ -265,52 +272,287 @@ async function fetchCurrentTasks(
...
+  } else {
+    // Auth middleware verified the agent's bearer token, so the row was
+    // present a moment ago. A vanishing row mid-heartbeat means it was
+    // deleted concurrently — surface it so an operator can correlate
+    // bursts of "ghost heartbeat" warnings with cleanup actions.
+    logger.warn(
+      { agentId, status: transition.effectiveStatus },
+      'Heartbeat for an agent row that no longer exists'
+    );
   }
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts:605]: > The description appears to be stale relative to the code that landed; updating it (or the residual list) would prevent reviewer/operator confusion about what is actually outstanding.

Addressed: the PR body was updated in commit 4c94b28 to remove the Remaining Residual Review Findings table entirely once all residuals (including #3, #6, and #7) landed. The current PR description reflects the actual state of the implementation.

File: packages/backend/src/services/agents.ts [hunk 9]
```diff
@@ -265,52 +272,287 @@ async function fetchCurrentTasks(
...
+      } catch (err) {
+        failed += 1;
+        logger.error(
+          { err, agentId, taskId: activeTask.id },
+          'handleTaskFailure threw during fatal-heartbeat fan-out; sibling tasks continue'
+        );
+      }
     }
+    taskFailureSummary = { attempted: activeTasks.length, failed };
   }
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts:636]: > consider hoisting `handleTaskFailure` to a top-level import (or caching the resolved module) so the import resolution is paid once.

Addressed in commit e5074af: added a module-level lazy cache (`getHandleTaskFailure`) so `await import('./tasks.js')` only resolves on the first fatal heartbeat; every subsequent fatal call uses the cached function reference. The dynamic import itself must stay (services/tasks.ts imports `getAgentBenchmarkForMode` from services/agents.ts, so a static top-level import would re-introduce the documented circular dependency). The post-commit positioning of the fan-out is intentional — handleTaskFailure runs its own transactional retry logic, and nesting drizzle transactions on the same connection produces savepoint churn we don't need; the residual race against concurrent task reassignment is bounded by the heartbeat-monitor sweep.

File: packages/backend/src/services/agents.ts [hunk 10]
```diff
@@ -265,52 +272,287 @@ async function fetchCurrentTasks(
...
+  const { updated, transition } = txResult;
+
+  // Post-commit emits + audit log. If the transaction rolled back,
+  // none of these fire and SSE listeners stay consistent.
   if (updated) {
-    emitAgentStatus(updated.projectId, updated.id, effectiveStatus);
+    if (data.error) {
+      emitAgentError(updated.projectId, updated.id, data.error.severity);
+    }
+    emitAgentStatus(updated.projectId, updated.id, transition.effectiveStatus);
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts:585]: > Consider whether warning-severity heartbeat errors should be coalesced or rate-limited at the emit layer

The existing per-(eventType, projectId) 250ms throttle in services/events.ts already coalesces bursts at the SSE layer (a busy agent emitting warnings every ~30s heartbeat is well under that window in normal operation; only thermal-spike loops at sub-second cadence would be affected). The agent contract is one heartbeat per ~30s, so flooding requires a misbehaving agent rather than a normal one — bounding warnings further would require event-type-aware coalescing that's wider than this PR. Tracking as a follow-up if real dashboards show warning-spam dominating the stream.

File: packages/backend/src/services/agents.ts [hunk 11]
```diff
@@ -265,52 +272,287 @@ async function fetchCurrentTasks(
...
+    'Heartbeat referenced currentTask.taskId not owned by this agent; dropping task linkage'
+  );
+  return undefined;
+}
+
+export async function processHeartbeat(agentId: number, data: AgentHeartbeat) {
+  // Verify the agent owns currentTask.taskId before the transaction
+  // opens. Done outside the tx so the read-only check can use the main
+  // pool and not extend lock duration on the agents row below.
+  const ownedTaskId = await verifyTaskOwnership(agentId, data.currentTask?.taskId);
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts]: > consider re-running the ownership check inside the transaction (or moving the SELECT into a CTE with the INSERT) to fully close the window.

Addressed in commit e5074af: `verifyTaskOwnership` now accepts a drizzle client and is called from inside the heartbeat transaction with `FOR UPDATE` on the task row. A concurrent reassignment is blocked on the lock until the heartbeat tx commits, so the task ownership at INSERT time matches the verification result.

File: packages/backend/src/services/agents.ts [hunk 12]
```diff
@@ -265,52 +272,287 @@ async function fetchCurrentTasks(
...
+/**
+ * Keys whose values should never reach `agent_errors.context`. An agent
+ * may inadvertently serialize `process.env`, HTTP headers, or
+ * exception-augmented `cause` chains that carry tokens/passwords;
+ * scrub them server-side so operators with dashboard access cannot
+ * read secrets that should have stayed on the agent host. The match
+ * is case-insensitive and substring-based so `API_KEY`, `apiKey`,
+ * `x-auth-token`, and `customer_secret` all redact.
+ */
+const SECRET_KEY_PATTERN = /token|password|secret|api[_-]?key|authorization|cookie|bearer/i;
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/services/agents.ts]: > The `SECRET_KEY_PATTERN` is substring-based on the key name. This will match legitimate non-secret keys whose names happen to contain one of the tokens

Addressed in commit e5074af: replaced the substring regex with a whole-word + trailing-suffix check (`isSecretKey`). Keys are normalized to snake_case (camelCase boundaries split, hyphens collapsed) and matched against a set of known secret names or as the trailing word of a compound. Now `tokenCount`, `cookieDomain`, `bearerHostname`, `apiKeyName`, `secretsManagerEnabled`, and `passwordAge` all pass through unchanged, while `api_key`, `apiKey`, `x-auth-token`, `customer_secret`, `db_password`, and `access_token` are still redacted.

File: packages/backend/tests/unit/agents-service.test.ts [hunk 13]
```diff
@@ -267,3 +269,209 @@ describe('AGENT_TASK_ACTIVE_STATUSES', () => {
...
+    expect(() => scrubAgentErrorContext(deep)).not.toThrow();
+  });
+
+  it('returns primitives unchanged', () => {
+    // Arrange + Act + Assert
+    expect(scrubAgentErrorContext('hello')).toBe('hello');
+    expect(scrubAgentErrorContext(42)).toBe(42);
+    expect(scrubAgentErrorContext(null)).toBeNull();
+    expect(scrubAgentErrorContext(undefined)).toBeUndefined();
+  });
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/unit/agents-service.test.ts:529]: > there is no negative test pinning the boundary (e.g., that `tokenCount` or `cookieDomain` are NOT redacted, or alternately that they ARE — whichever the policy is). Add a test that documents the intended behavior

Addressed in commit e5074af: added two tests in agents-service.test.ts — `'preserves descriptive keys that only contain a secret-name as a prefix'` pins the false-positive-avoidance behavior (`tokenCount`, `cookieDomain`, `bearerHostname`, `apiKeyName`, `secretsManagerEnabled`, `passwordAge` pass through), and `'redacts trailing-secret compound keys (e.g., customer_secret, db_password)\'` pins the positive case for compound names (`customer_secret`, `db_password`, `access_token`, `refresh_token`, `x_api_key`). Future tightening of the regex now has clear failure modes.

File: packages/shared/src/db/migrations/0007_light_ulik.sql [hunk 14]
```diff
@@ -0,0 +1,11 @@
+-- Clear any orphan task references before attaching the FK. Pre-existing
+-- agent_errors rows may reference task ids that have since been deleted;
+-- ON DELETE SET NULL only applies to deletes that happen AFTER the FK is
+-- in place, so we have to backfill old orphans manually.
+UPDATE "agent_errors"
+SET "task_id" = NULL
+WHERE "task_id" IS NOT NULL
+  AND "task_id" NOT IN (SELECT "id" FROM "tasks");--> statement-breakpoint
+DROP INDEX "agent_errors_agent_id_created_at_idx";--> statement-breakpoint
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/shared/src/db/migrations/0007_light_ulik.sql]: > The DROP INDEX has no `IF EXISTS` clause — if any environment was previously hand-rebuilt or skipped a prior migration, the DROP will fail and block the FK addition.

Addressed in commit e5074af: added `IF EXISTS` to the DROP. A comment in the migration documents why the recreate happens (Drizzle's NULLS-clause normalization — pg's implicit NULLS FIRST on DESC vs. the new explicit NULLS LAST). The brief no-index window between DROP and CREATE is acceptable because both statements run in the same drizzle invocation; `CREATE INDEX CONCURRENTLY` was considered but it cannot run inside the migration's transactional wrapper, so a follow-up would need to disable the wrapper for that one statement.

File: packages/backend/src/middleware/auth.ts [hunk 15]
```diff
@@ -50,36 +50,67 @@ export const requireSession = createMiddleware<AppEnv>(async (c, next) => {
...
+/**
+ * Recovery-friendly agent auth — accepts agents whose row is in
+ * `status='error'` so they can post a recovery heartbeat. Use ONLY on
+ * `POST /api/v1/agent/heartbeat`; the heartbeat handler is responsible
+ * for transitioning the agent back to `online` (or keeping it in
+ * `error` if the recovery heartbeat itself reports a fatal error).
+ */
+export const requireAgentTokenAllowRecovery = createAgentTokenMiddleware({
+  allowErroredAgent: true,
 });
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/src/middleware/auth.ts:116]: > Consider either (a) adding a method/path assertion inside the middleware to enforce it can only attach to heartbeat, or (b) renaming to something like `requireAgentTokenForHeartbeatRecovery` so the mistake is harder to make.

Addressed in commit e5074af: renamed to `requireAgentTokenForHeartbeatRecovery` (option b). All call sites, tests, and the docstring updated to match. Future re-use on another endpoint now requires the author to override an explicitly heartbeat-named symbol, which is the friction this rename was meant to introduce.

File: packages/backend/tests/integration/agent-heartbeat.test.ts [hunk 16]
```diff
@@ -0,0 +1,529 @@
...
+            };
+          }
+        }
+        return buildSelectChain(state.agent ? [state.agent] : []);
+      },
+      insert: (_table: AnyTable) => buildAgentErrorsInsertChain(),
+      update: (_table: AnyTable) => buildAgentsUpdateChain(),
+      delete: (_table: AnyTable) => ({ where: () => Promise.resolve() }),
+    };
+  }
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/integration/agent-heartbeat.test.ts:217]: > Consider either (a) gating these as real-DB integration tests against the test Postgres instance the repo already uses, or (b) making the mock match on the constructed SQL/where clause rather than projection-key shape. The current mock is workable as a stopgap but is a long-term maintenance burden.

Acknowledged. The disambiguation shape is documented in an inline comment that lists the three call sites (verifyTaskOwnership, active-task fan-out, high-priority lookup) so a future refactor changing the SELECT shapes will surface as a test failure with a clear pointer. Promoting these to real-DB tests requires the project to land a tests/integration/ Postgres harness first — that infrastructure work is outside this PR's scope, but the file structure (isolated phase, scoped state reset) is intentionally compatible with a future swap to real-DB.

File: packages/backend/tests/integration/agent-heartbeat.test.ts [hunk 17]
```diff
@@ -0,0 +1,529 @@
...
+
+  beforeEach(() => {
+    resetState();
+    handleTaskFailureMock.mockClear();
+    loggerMock.info.mockClear();
+    loggerMock.warn.mockClear();
+    loggerMock.error.mockClear();
+    emitAgentErrorMock.mockClear();
+    emitAgentStatusMock.mockClear();
+  });
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/backend/tests/integration/agent-heartbeat.test.ts:296]: > Use `mockReset()` instead of `mockClear()` to prevent queued value leakage.

Addressed in commit e5074af: `beforeEach` in tests/integration/agent-heartbeat.test.ts now uses `mockReset()` on every mock and reinstalls the default `handleTaskFailure` implementation (extracted as `defaultHandleTaskFailureImpl`) so the sibling-failure test's `mockImplementationOnce` queue can't leak across cases.

File: packages/shared/src/db/migrations/meta/_journal.json [hunk 18]
```diff
@@ -50,6 +50,13 @@
       "when": 1779066400000,
       "tag": "0006_agent_errors_composite_index",
       "breakpoints": true
+    },
+    {
+      "idx": 7,
+      "version": "7",
+      "when": 1779052939557,
+      "tag": "0007_light_ulik",
+      "breakpoints": true
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply packages/shared/src/db/migrations/meta/_journal.json:59]: > Regenerate migration 0007 with correct timestamp ordering.

Addressed in commit e5074af: updated the `when` value for migration 0007 in `_journal.json` from 1779052939557 to 1779152939557, which is approximately 24 hours after migration 0006 (1779066400000). Chronological order is now strictly increasing across all 8 migrations.
